mountpoint-s3/src/inode.rs

```diff
@@ -130,7 +130,20 @@ impl Superblock {
                 if new_lookup_count == 0 {
                     // Safe to remove, kernel no longer has a reference to it.
                     trace!(ino, "removing inode from superblock");
-                    inodes.remove(&ino);
+                    let inode = inodes.remove(&ino).expect("we just retrieved it");
+
+                    // Should be impossible for this to fail (VFS inodes reference their parent, so
+                    // children need to be freed first), but let's not crash in a `forget` function...
+                    let Some(parent) = inodes.get(&inode.parent()) else {
+                        debug_assert!(false, "children should be forgotten before parents");
+                        return;
+                    };
+                    let mut parent_state = parent.inner.sync.write().unwrap();
+                    let InodeKindData::Directory { children, writing_children, .. } = &mut parent_state.kind_data else {
+                        unreachable!("parent is always a directory");
+                    };
+                    children.remove(inode.name());
```

Comment:
Should we check that we are in fact removing the same inode? vs a new child with the same name?

Reply:
It should be possible to write a test for this scenario, e.g.:
* unlink child
* mknod new child with same name
* forget original child

Reply:
Nice catch. I wrote a slightly different test with the same idea -- the child changes from a file to a directory and so gets recreated, forgetting the original child.